cxplat/src/cxplat_winuser/workitem_winuser.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "cxplat.h"
+#include "winuser_internal.h"
+#include <windows.h>
+
+// Thread pool related globals.
+static TP_CALLBACK_ENVIRON _callback_environment{};
+static PTP_POOL _pool = nullptr;
+static PTP_CLEANUP_GROUP _cleanup_group = nullptr;
+
+typedef struct _cxplat_preemptible_work_item
+{
+    PTP_WORK work;
+    void (*work_item_routine)(_Inout_opt_ void* work_item_context);
+    void* work_item_context;
+} cxplat_preemptible_work_item_t;
+
+static cxplat_rundown_reference_t _cxplat_preemptible_work_items_rundown_reference;
+
+static void
+_cxplat_preemptible_routine(_Inout_ PTP_CALLBACK_INSTANCE instance, _In_opt_ void* parameter, _Inout_ PTP_WORK work)
+{
+    UNREFERENCED_PARAMETER(instance);
+    UNREFERENCED_PARAMETER(work);
+
+    if (parameter == nullptr) {
+        return;
+    }
+
+    cxplat_preemptible_work_item_t* work_item = (cxplat_preemptible_work_item_t*)parameter;
+    work_item->work_item_routine(work_item->work_item_context);
+
+    cxplat_free_preemptible_work_item(work_item);
+}
+
+_Must_inspect_result_ cxplat_status_t
+cxplat_allocate_preemptible_work_item(
+    _In_opt_ void* caller_context,
+    _Outptr_ cxplat_preemptible_work_item_t** work_item,
+    _In_ void (*work_item_routine)(_Inout_opt_ void* work_item_context),
+    _Inout_opt_ void* work_item_context)
+{
+    UNREFERENCED_PARAMETER(caller_context);
+    cxplat_status_t result = CXPLAT_STATUS_SUCCESS;
+
+    if (!cxplat_acquire_rundown_protection(&_cxplat_preemptible_work_items_rundown_reference)) {
+        return CXPLAT_STATUS_UNSUCCESSFUL;
+    }
+
+    *work_item = (cxplat_preemptible_work_item_t*)cxplat_allocate(sizeof(cxplat_preemptible_work_item_t));
+    if (*work_item == nullptr) {
+        result = CXPLAT_STATUS_NO_MEMORY;
+        goto Done;
+    }
+
+    // It is required to use the InitializeThreadpoolEnvironment function to
+    // initialize the _callback_environment structure before calling CreateThreadpoolWork.
+    (*work_item)->work = CreateThreadpoolWork(_cxplat_preemptible_routine, *work_item, &_callback_environment);
+    if ((*work_item)->work == nullptr) {
+        result = CXPLAT_STATUS_FROM_WIN32(GetLastError());
```

Comment:
We're going to need to add logging eventually at every first point of failure, like this.

Reply:
Agree